### PR TITLE
[clang-tidy] Add check `readability-avoid-return-with-void-value`

### LINKED ISSUE #75788: Readability check to detect `return` from a `void` function
Hi!

We recently came across this type of code and realized that some tool should have caught it:

```cpp
void foo()
{}

void bar()
{
    return foo();
}
```

Turns out no warning in clang or clang-tidy catches this. Maybe we could have a clang-tidy `readability` check for it?

### PR DESCRIPTION
**Pull request by SimplyDanny**
Closes #75788.

The idea is to check each return statements for an expression that has type `void` as this is only possible in a function with `void` return type.

The implementation seems too simple. I might have missed something.

### CODE REVIEW

File: clang-tools-extra/clang-tidy/readability/ReadabilityTidyModule.cpp [hunk 1]
```diff
@@ -119,6 +120,8 @@ class ReadabilityModule : public ClangTidyModule {
         "readability-redundant-preprocessor");
     CheckFactories.registerCheck<ReferenceToConstructedTemporaryCheck>(
         "readability-reference-to-constructed-temporary");
+    CheckFactories.registerCheck<ReturnExpressionInVoidFunctionCheck>(
+        "readability-return-expression-in-void-function");
```
Comment (PiotrZSL) [inline clang-tools-extra/clang-tidy/readability/ReadabilityTidyModule.cpp:124]: I'm not sure about this name, if this is about return argument, or just an return

File: clang-tools-extra/clang-tidy/readability/ReturnExpressionInVoidFunctionCheck.cpp [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    MatchFinder *Finder) {
+  Finder->addMatcher(
+      returnStmt(hasReturnValue(hasType(voidType()))).bind("void_return"),
+      this);
+}
+
+void ReturnExpressionInVoidFunctionCheck::check(
+    const MatchFinder::MatchResult &Result) {
+  const auto *VoidReturn = Result.Nodes.getNodeAs<ReturnStmt>("void_return");
+  diag(VoidReturn->getBeginLoc(), "return statements should not return void");
```
Comment (PiotrZSL) [inline clang-tools-extra/clang-tidy/readability/ReturnExpressionInVoidFunctionCheck.cpp:28]: message could be better, to point out that that it's about returning void expression not, void type,
maybe better that return statement in void function shouldn't have subexpression/value

File: clang-tools-extra/clang-tidy/readability/ReturnExpressionInVoidFunctionCheck.h [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+/// Looks for statements returning expressions of type `void`.
+///
+/// For the user-facing documentation see:
+/// http://clang.llvm.org/extra/clang-tidy/checks/readability/return-expression-in-void-function.html
+class ReturnExpressionInVoidFunctionCheck : public ClangTidyCheck {
+public:
+  ReturnExpressionInVoidFunctionCheck(StringRef Name, ClangTidyContext *Context)
+      : ClangTidyCheck(Name, Context) {}
+  void registerMatchers(ast_matchers::MatchFinder *Finder) override;
+  void check(const ast_matchers::MatchFinder::MatchResult &Result) override;
```
Comment (PiotrZSL) [inline clang-tools-extra/clang-tidy/readability/ReturnExpressionInVoidFunctionCheck.h:25]: ignore implicit code with TK_IgnoreUnlessSpelledInSource

File: clang-tools-extra/docs/clang-tidy/checks/list.rst [hunk 4]
```diff
@@ -371,6 +371,7 @@ Clang-Tidy Checks
    :doc:`readability-redundant-string-cstr <readability/redundant-string-cstr>`, "Yes"
    :doc:`readability-redundant-string-init <readability/redundant-string-init>`, "Yes"
    :doc:`readability-reference-to-constructed-temporary <readability/reference-to-constructed-temporary>`,
+   :doc:`readability-return-expression-in-void-function <readability/return-expression-in-void-function>`, "Yes"
```
Comment (PiotrZSL) [inline clang-tools-extra/docs/clang-tidy/checks/list.rst:374]: remove "Yes", as no fixes are provided

**Review by PiotrZSL — COMMENTED** [review]
Not too bad, think more about check name and warning message.

File: clang-tools-extra/test/clang-tidy/checkers/readability/return-expression-in-void-function.cpp [hunk 5]
```diff
@@ -0,0 +1,23 @@
+// RUN: %check_clang_tidy %s readability-return-expression-in-void-function %t
+
+void f1();
+
+void f2() {
+    return f1();
```
Comment (PiotrZSL) [inline clang-tools-extra/test/clang-tidy/checkers/readability/return-expression-in-void-function.cpp:6]: add tests with normal returns: "return;"

File: clang-tools-extra/clang-tidy/readability/ReadabilityTidyModule.cpp (hunk 1 above)
Comment (SimplyDanny) [reply clang-tools-extra/clang-tidy/readability/ReadabilityTidyModule.cpp:124]: What about the succinct `readability-void-return`?

Comment (PiotrZSL) [reply clang-tools-extra/clang-tidy/readability/ReadabilityTidyModule.cpp:124]: Consider: readability-return-void-value or readability-void-return-statement or readability-avoid-return-with-void-value.

Comment (SimplyDanny) [reply clang-tools-extra/clang-tidy/readability/ReadabilityTidyModule.cpp]: `readability-avoid-return-with-void-value` sounds good.

File: clang-tools-extra/clang-tidy/readability/AvoidReturnWithVoidValueCheck.cpp [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+  Finder->addMatcher(
+      returnStmt(hasReturnValue(hasType(voidType()))).bind("void_return"),
+      this);
+}
+
+void AvoidReturnWithVoidValueCheck::check(
+    const MatchFinder::MatchResult &Result) {
+  const auto *VoidReturn = Result.Nodes.getNodeAs<ReturnStmt>("void_return");
+  diag(VoidReturn->getBeginLoc(),
+       "return statement in void function should not return a value");
```
Comment (PiotrZSL) [inline clang-tools-extra/clang-tidy/readability/AvoidReturnWithVoidValueCheck.cpp:28]: consider "return statement within a void function should not have a specified return value"

File: clang-tools-extra/clang-tidy/readability/AvoidReturnWithVoidValueCheck.cpp [hunk 7]
```diff
@@ -0,0 +1,31 @@
+//===--- AvoidReturnWithVoidValueCheck.cpp - clang-tidy -------------===//
```
Comment (PiotrZSL) [inline clang-tools-extra/clang-tidy/readability/AvoidReturnWithVoidValueCheck.cpp:1]: format this line to match length of line 7

File: clang-tools-extra/docs/ReleaseNotes.rst [hunk 8]
```diff
@@ -218,6 +218,13 @@ New checks
   Detects C++ code where a reference variable is used to extend the lifetime
   of a temporary object that has just been constructed.
 
+- New :doc:`readability-avoid-return-with-void-value
+  <clang-tidy/checks/readability/avoid-return-with-void-value>` check.
+
+  Complains about statements returning expressions of type ``void``. It can be
```
Comment (PiotrZSL) [inline clang-tools-extra/docs/ReleaseNotes.rst:224]: Consider simplifying description of check make to something like "Finds return statements with void values used within functions with a void result type."

File: clang-tools-extra/docs/clang-tidy/checks/readability/avoid-return-with-void-value.rst [hunk 9]
```diff
@@ -0,0 +1,30 @@
+.. title:: clang-tidy - readability-avoid-return-with-void-value
+
+readability-avoid-return-with-void-value
+========================================
+
+Complains about statements returning expressions of type ``void``. It can be
+confusing if a function returns an expression even though its return type is
+``void``.
```
Comment (PiotrZSL) [inline clang-tools-extra/docs/clang-tidy/checks/readability/avoid-return-with-void-value.rst:8]: consider: "Void function is intended to perform a task without producing a return value. Including a return value in such functions can lead to confusion and may miscommunicate the function's intended behavior." or something similar.

File: clang-tools-extra/clang-tidy/readability/AvoidReturnWithVoidValueCheck.h [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+/// http://clang.llvm.org/extra/clang-tidy/checks/readability/avoid-return-with-void-value.html
+class AvoidReturnWithVoidValueCheck : public ClangTidyCheck {
+public:
+  AvoidReturnWithVoidValueCheck(StringRef Name, ClangTidyContext *Context)
+      : ClangTidyCheck(Name, Context) {}
+  void registerMatchers(ast_matchers::MatchFinder *Finder) override;
+  void check(const ast_matchers::MatchFinder::MatchResult &Result) override;
+  std::optional<TraversalKind> getCheckTraversalKind() const override {
+    return TK_IgnoreUnlessSpelledInSource;
+  }
```
Comment (PiotrZSL) [inline clang-tools-extra/clang-tidy/readability/AvoidReturnWithVoidValueCheck.h:28]: Limit check to C++ only, look to other checks how to do that.

File: clang-tools-extra/docs/clang-tidy/checks/readability/avoid-return-with-void-value.rst [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+
```
Comment (PiotrZSL) [inline clang-tools-extra/docs/clang-tidy/checks/readability/avoid-return-with-void-value.rst:9]: Add info that in "C" those issues are detected by -Wpedantic warning.
`warning: void function 'f' should not return void expression [-Wpedantic]`

**Review by PiotrZSL — COMMENTED** [review]
Last nits.

File: clang-tools-extra/docs/ReleaseNotes.rst [hunk 12]
```diff
@@ -221,9 +221,8 @@ New checks
 - New :doc:`readability-avoid-return-with-void-value
   <clang-tidy/checks/readability/avoid-return-with-void-value>` check.
 
-  Complains about statements returning expressions of type ``void``. It can be
-  confusing if a function returns an expression even though its return type is
-  ``void``.
+  Finds return statements with ``void`` values used within functions with
```
Comment (PiotrZSL) [inline clang-tools-extra/docs/ReleaseNotes.rst:224]: synchronize class description, release note and first sentence in documentation.

**Comment by PiotrZSL** [comment]
One more thing, there are issues like this reported in gtest, soe maybe would be good to add IgnoreMacros options like in other checks.

**Comment by PiotrZSL** [comment]
I run check on llvm-project, finds thousands of issues. Most of them comes from google test but other issues are like this one:
```
if (x)
  return call();
```

Just some simplification to not write {}.
I think that both IgnoreMacros and StrictMode(false) could be needed. If StrictMode is false then only returnStmt that got compoundStmt as parent should be check, to reduce impact.

**Comment by SimplyDanny** [comment]
> I think that both IgnoreMacros and StrictMode(false) could be needed. If StrictMode is false then only returnStmt that got compoundStmt as parent should be check, to reduce impact.

Fair points. Both options are added now. However, I'm not sure about the name of the second one `StrictMode`. It doesn't imply its purpose. Do you think we should pick a more descriptive name?

I also found a case involving initializer lists that requires special treatment, because the list is seen as an expression returning `void`. This is addressed in the [last commit](https://github.com/llvm/llvm-project/pull/76249/commits/ce006f2957402a8a7bacc023cc96963b211e1290).

**Comment by PiotrZSL** [comment]
@SimplyDanny If you want other name than StrictMode then you can always something like IgnoreNonSplittableReturns or IgnoreSingleStatementBlocks, or SafeMode. Idea about StrictMode is that many checks use that, when set to true, then check should be more strict and detect more issues, with false check should be less strict and detect less issues.

File: clang-tools-extra/docs/clang-tidy/checks/readability/avoid-return-with-void-value.rst [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+In C, the same issue is detected by the compiler if the ``-Wpedantic`` mode
+is enabled.
+
+Options
+-------
+
+.. option::  IgnoreMacros
+
+  The value ``false`` specifies that return statements expanded
```
Comment (PiotrZSL) [inline clang-tools-extra/docs/clang-tidy/checks/readability/avoid-return-with-void-value.rst:43]: options values should be in single \`, so change \`\`false\`\` into \`false\`, same for other option and true value

**Review by PiotrZSL — APPROVED** [review]
Overall looks fine to me.
Just few minor nits.

**Comment by SimplyDanny** [comment]
Thank you, @PiotrZSL, for the reviews and helpful suggestions!

**Pull request merged**